docs(core): primary Angular Signals documentation (#50053)
angular/angular: aio/content/guide/signals.md

# Angular Signals

> **Signals** are used to declare and update the data displayed in templates in a way that Angular can track and optimize.

Reviewer: +1, I like the shifted focus to a more holistic framing.

<div class="alert is-important">

> Angular signals are available for [developer preview](https://angular.io/guide/releases#developer-preview). They're ready for you to try, but may change before they are stable.

Reviewer 2: We can keep the link relative :
```suggestion
Angular signals are available for [developer preview](/guide/releases#developer-preview). They're ready for you to try, but may change before they are stable.
```

</div>

## What are signals?

A signal is a wrapper around a value, which is capable of notifying interested consumers when that value changes. There are different kinds of signals.

> Some signals can have their value directly updated via a mutation API. In our implementation, these are known as "writable" signals. Updates to the model are always done by changing one or more writable signals.

Reviewer: I think you can strike the `mutation API` part -- it's clearer to just talk about `writable signals` in the introduction here.

Because reading a signal is done through a getter, signals are able to keep track of where they're being read. And because mutations are always done with the mutation API, signals know when they're mutated and inform previous readers about the change.

## Writable signals

Writable signals are created via the `signal` function, and always start with an initial value:

```ts
const count = signal(0);

// Signals are getter functions - calling them reads their value.
console.log('The count is: ' + count());
```

To change the value of a writable signal, you can either `.set()` it directly:

```ts
count.set(3);
```

or use the `.update()` operation to compute a new value from the previous one:

```ts
// Increment the count by 1.
count.update(value => value + 1);
```

> Values inside writable signals can be simple primitives or complex nested objects. In the latter case, sometimes we want to mutate those objects without needing to produce entirely new values, such as pushing new data to an array. To do that, we can use the `.mutate` operation:

Reviewer 3: ```suggestion
If your signals contains an object, you may want to mutate that object directly without creating a new instance. One common example is pushing a new value into an array. To update an object, use the `.mutate` method:
```
I grabbed the first sentence here and snuck a slightly modified version of it into the introductory paragraph on signals since it seems more foundational. 

We'd generally also want to avoid first-person in docs ("I" and "we"), preferring "you". Even better, though, if we can make the "you" implicit an imperative.

```ts
const todos = signal([{title: 'Learn signals', done: false}]);

todos.mutate(value => {

>   // make a change to the value of the signal without replacing it

Reviewer 3: ```suggestion
  // Change the value of the signal without replacing it
```

  value[0].done = true;
});
```

## Computed signals

> A computed signal has a value which cannot be changed directly, but which is derived from the values of other signals.
> 

Reviewer 3: ```suggestion
A **computed signal** derives its value from other signals. You cannot directly assign values to a computed signal. 

```
I would flip the order here, since the derivation is the important part and the non-assignability is a consequence of that derivation.

Going further, though, I think it might be best to move the latter part to its own paragraph at the end of this section, since it feels slightly out of flow with definition -> example.

```typescript
const count: WritableSignal<number> = signal(0);
const doubleCount: Signal<number> = computed(() => count * 2);
```

> The `doubleCount` signal has a dependency on `count`. Whenever `count` updates, `doubleCount` will be notified that it needs to recompute. A `computed` will rerun when its dependencies update and it memoizes (caches) the result. Reading the value from a computed multiple times does not cause the computation to re-execute unless a dependency has changed since the last run.

Reviewer 3: ```suggestion
The `doubleCount` signal depends on `count`. Whenever `count` updates, it tells `doubleCount` that it needs to recompute. Each `computed` reruns when its dependencies update and it memoizes (caches) the result. Reading the value from a computed multiple times does not cause the computation to re-execute unless a dependency has changed since the last run.
```
suggestions here for: tense (always present), passive voice, some brevity

### `Computed` is lazy

> The body of a `computed` does not execute until it is read for the first time. In addition, only the signals read during the computation body are tracked. For example:

Reviewer 3: ```suggestion
The function passed to `computed` does not run until something reads the computed signal for the first time. When run, the computed signal tracks only the signals that it reads.
```

```ts
const showCount = signal(false);
const count = signal(0);
const conditionalCount = computed(() => {
  if (showCount()) {
    return `The count is ${count()}.`;
  } else {
    return 'Nothing to see here!';
  }
});
```

> `conditionalCount` signal does not execute its computation function until it's read for the first time.

Reviewer 3: ```suggestion
`conditionalCount` does not execute its computation function until it's read for the first time.
```

If we read `conditionalCount` and `showCount` is `false`, we return the "Nothing to see here!" message _without_ reading the `count` signal. This means that updates to `count` will not result in a recomputation.

If we then update `showCount` to be `true` and read `conditionalCount`, the computation will re-execute, find that `showCount` is `true`, and take the branch which returns "The count is" with the value of `count`. Changes to `count` will then invalidate `conditionalCount`'s cached value. Note that if `showCount` is later set to `false` again, then `count` will no longer be considered a dependency of `conditionalCount`.

> 

Reviewer: I would conclude this section by reiterating the main point, just because it's so important, and people might skim the example. Something like:

> In summary, this example shows how a computed only updates based on a signal if it *actually* read that signal during its previous update.

You might be able to write a clearer summary than this, but it's a bit tricky to get down to one sentence.

## Reading signals in `OnPush` components

> When a signal is read in the template of an `OnPush` component, that component becomes a tracked consumer of the signal. When that signal is updated and notifies its consumers, Angular will automatically mark the component for check and it will be updated the next time change detection runs. Refer to the [Skipping component subtrees](https://angular.io/guide/change-detection-skipping-subtrees) guide for more information about `OnPush` components.

Reviewer 2: Same as above 👍 
```suggestion
When a signal is read in the template of an `OnPush` component, that component becomes a tracked consumer of the signal. When that signal is updated and notifies its consumers, Angular will automatically mark the component for check and it will be updated the next time change detection runs. Refer to the [Skipping component subtrees](/guide/change-detection-skipping-subtrees) guide for more information about `OnPush` components.
```

Author: I did a little poking around other places where we reference e.g. the developer preview guide, and they all use an absolute link.

## Effects

> Signals are useful because they can notify interested consumers when they change. An operation that's performed when the signals it depends on are changed is called an "effect". For example, Angular uses an effect to mark an `OnPush` component for check whenever any signals used within that component's template are changed.
> 
> Using the `effect` function, you can create your own effects:

Reviewer 3: ```suggestion
Signals are useful because they can notify interested consumers when they change. An **effect** is an operation that runs whenever one or more signal values change. You can create an effect with the `effect` function.
```

I would move the part here about `OnPush` to later in this section, as something like 
```markdown
Internally, Angular uses `effects` for its own behaviors. For example, the framework uses
an effect to mark `OnPush` components for change detection whenever any signal values
read within that component's template change.
```

Author: I disagree about moving the `OnPush` section.

Using signals to mark `OnPush` components for check is the primary use case (in v16), which doesn't require understanding or creating effects.

```ts
effect(() => {
  console.log(`The current count is: ${count()}`);
});
```

> Effects execute once initially, and then whenever one or more of the signals they read are updated. Like with computed signals, effects keep track of their dependencies dynamically, so signals inside conditionals that don't execute are not considered dependencies.

Reviewer 3: ```suggestion
**Effects always run at least once.** During this first run, the effect tracks any signal value reads. Whenever any of these signal values change, the effect runs again. Similar to computed signals, effects keep track of their dependencies dynamically, so the effect does not track any signal values that are not read.
```

> Effects always execute asynchronously, before and during change detection.

Reviewer 3: ```suggestion
**Effects always execute asynchronously.** <something about timing here>
```

I don't quite follow what's meant by "before and during change detection" here.

> ### Uses for effects

Reviewer 3: ```suggestion
### Best practices for `effect`
```
I might combine these two sections into one best practices section

> Effects are rarely needed in most application code, but may be useful in specific circumstances. Here are some examples of situations where an `effect` might be a good solution:

Reviewer 3: ```suggestion
Use `effect` only when you must directly call an explicit API based on
state changes. Some common examples of good cases for `effect` include:
```

* Logging data being displayed and when it changes, either for analytics or as a debugging tool
* Keeping data in sync with `window.localStorage`

> * Adding custom DOM behavior that can't be expressed with template syntax

Reviewer 3: ```suggestion
* Adding custom DOM behavior that cannot be expressed with template syntax
```

* Performing custom rendering to a `<canvas>`, charting library, or other third party UI library

#### When not to use effects

> Avoid using effects for propagation of state changes. This can result in `ExpressionChangedAfterItHasBeenChecked` errors, infinite circular updates, or unnecessary change detection cycles.

Reviewer 3: ```suggestion
Avoid using `effect` to propagate state changes. This can result in `ExpressionChangedAfterItHasBeenChecked` errors, infinite circular updates, or unnecessary change detection cycles.
```

Because of these risks, setting signals is disallowed by default in effects, but can be enabled if absolutely necessary.

### Injection context

> By default, registering a new effect with the `effect()` function requires an "injection context" (access to the `inject` function). The easiest way to provide this is to call `effect` within a component, directive, or service `constructor`:

Reviewer 3: ```suggestion
By default, registering a new effect with the `effect()` function requires an _injection context_ (access to the `inject` function). The easiest way to provide this is to call `effect` within a component, directive, or service `constructor`:
```

```ts
@Component({...})
export class EffectiveCounterCmp {
  readonly count = signal(0);
  constructor() {
    // Register a new effect.
    effect(() => {
      console.log(`The count is: ${this.count()})`);
    });
  }
}
```

> Alternatively, the effect can be assigned to a field (which also gives it a descriptive name).

Reviewer 3: ```suggestion
Alternatively, the effect can be assigned to a class property, which also gives it a descriptive name.
```
I might use "class property" in case "field" is perhaps more esoteric

```ts
@Component({...})
export class EffectiveCounterCmp {
  readonly count = signal(0);
  
  private loggingEffect = effect(() => {
    console.log(`The count is: ${this.count()})`);
  });
}
```

To create an effect outside of the constructor, you can pass an `Injector` to `effect` via its options:

```ts
@Component({...})
export class EffectiveCounterCmp {
  readonly count = signal(0);
  constructor(private injector: Injector) {}

  initializeLogging(): void {
    effect(() => {
      console.log(`The count is: ${this.count()})`);
    }, {injector: this.injector});
  }
}
```

### Destroying effects

> When you create an effect, it will automatically be destroyed when its enclosing context is destroyed. This means that effects created within components will be destroyed when the component is destroyed. The same goes for effects within directives, services, etc.
> 
> Effects return an `EffectRef` that can be used to destroy them manually, via the `.destroy()` operation. This can also be combined with the `manualCleanup` option to create an effect that will last until it is manually destroyed. Be careful to actually clean up such effects when it's no longer required.

Reviewer 3: ```suggestion
Effects are automatically destroyed when their enclosing context is destroyed. Effects created within components are destroyed when the component is destroyed. The same goes for effects within directives, services, etc.

Effects return an `EffectRef` that lets you destroy them manually, via the `.destroy()` function. You can combine this with the `manualCleanup` option to create an effect that lasts until it is manually destroyed. Always clean up such effects when it's no longer required.
```

## Advanced topics

### Signal equality functions

> When creating a signal, you can optionally provide an equality function, which will be used to check whether the new value is actually different than the previous one.

Reviewer 3: ```suggestion
When creating a signal, you can optionally provide an equality function, which is used to check whether the new value is different than the previous one.
```

```ts
import _ from 'lodash';

const data = signal(['test'], {equal: _.isEqual});

// Even though this is a different array instance, the deep equality
// function will consider the values to be equal, and the signal won't
// trigger any updates.
data.set(['test']);
```

> Equality functions can be provided to both writable and computed signals.

Reviewer 3: ```suggestion
You can set equality functions for both writable and computed signals.
```

> For writable signals, `.mutate()` always bypasses the equality check, because it mutates the current value and doesn't produce a new one.

Reviewer 3: ```suggestion
For writable signals, `.mutate()` always ignores the equality check because it mutates the current value without producing a new reference.
```

Author: Changed to "does not check for equality"

Effects return an `EffectRef` that can be used to destroy them manually, via the `.destroy()` operation. This can also be combined with the `manualCleanup` option to create an effect that will last until it is manually destroyed. Be careful to actually clean up such effects when it's no longer required.

## Advanced topics

### Signal equality functions

When creating a signal, you can optionally provide an equality function, which will be used to check whether the new value is actually different than the previous one.

```ts
import _ from 'lodash';

const data = signal(['test'], {equal: _.isEqual});

// Even though this is a different array instance, the deep equality
// function will consider the values to be equal, and the signal won't
// trigger any updates.
data.set(['test']);
```

Equality functions can be provided to both writable and computed signals.

For writable signals, `.mutate()` always bypasses the equality check, because it mutates the current value and doesn't produce a new one.

### Reading without tracking dependencies

> Occasionally it's necessary to execute code which might read signals in a reactive function (such as a computed signal or an effect) _without_ creating a dependency.

Reviewer 3: ```suggestion
Rarely, you may want to read signal values in a reactive function, such as `computed` or `effect`, without tracking a dependency on that value.
```

I suspect this would/should be rare?

For example, suppose that when `currentUser` changes, the value of a `counter` should be logged. Creating an `effect` which reads both signals:

```ts
effect(() => {
  console.log(`User set to `${currentUser()}` and the counter is ${counter()}`);
});
```

> will log the message when _either_ the `currentUser` or the `counter` changes. But the effect should only run when `currentUser` changes, it's only incidental that it reads the `counter` to produce the log message.

Reviewer 3: ```suggestion
This example logs a message when _either_ `currentUser` or `counter` changes. However, we want the effect to run only when `currentUser` changes because the `counter` is incidental.
```

> To ensure that the effect depends on `currentUser` but _not_ on `counter`, the effect can do an _untracked_ read of `counter`, using the `untracked` helper function. `untracked` will run a function and return its result without creating any dependencies:

Reviewer 3: ```suggestion
You can prevent a signal read from being tracked by calling its getter with `untracked`.
```

```ts
effect(() => {
  console.log(`User set to `${currentUser()}` and the counter is ${untracked(() => counter())}`);
});
```

Because a signal getter is a function, untracked reads can be simplified by passing the signal itself to `untracked`:

```ts
effect(() => {
  console.log(`User set to `${currentUser()}` and the counter is ${untracked(counter)}`);
});
```

`untracked` is also useful when an effect needs to invoke some external code which shouldn't be treated as a dependency:

```ts
effect(() => {
  const user = currentUser();
  untracked(() => {
    // If the `loggingService` reads signals, they won't be counted as
    // dependencies of this effect.
    this.loggingService.log(`User set to ${user}`);
  });
});
```

### Effect cleanup functions

> Effects might start long-running operations, which should be cancelled if the effect is destroyed or runs again before the first operation finished. To do this, you can register a cleanup function when the effect runs. That cleanup function will be called before the effect runs again, or if it's destroyed.

Reviewer 3: ```suggestion
Effects might start long-running operations, which should be cancelled if the effect is destroyed or runs again before the first operation finished. When you create an effect, your function can optionally accept an `onCleanup` function as its first parameter. This `onCleanup` function lets you register a callback that is invoked when the effect is destroyed.
```
